In [1]:
import pandas as pd
import numpy as np
import random
import re

In [2]:
data_url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/" \
            "IBMDeveloperSkillsNetwork-DS0103EN-SkillsNetwork/labs/Module%202/recipes.csv"

#### Load the data from the link
The dataset contains food recipes from various countries and their ingredients. The target variable is the name of the country the recipe is from.

In [3]:
recipes = pd.read_csv(data_url)

Show the first 5 items

In [4]:
recipes.head()

country almond angelica anise anise_seed apple apple_brandy apricot  \
0  Vietnamese     No       No    No         No    No           No      No   
1  Vietnamese     No       No    No         No    No           No      No   
2  Vietnamese     No       No    No         No    No           No      No   
3  Vietnamese     No       No    No         No    No           No      No   
4  Vietnamese     No       No    No         No    No           No      No   

  armagnac artemisia  ... whiskey white_bread white_wine  \
0       No        No  ...      No          No         No   
1       No        No  ...      No          No         No   
2       No        No  ...      No          No         No   
3       No        No  ...      No          No         No   
4       No        No  ...      No          No         No   

  whole_grain_wheat_flour wine wood yam yeast yogurt zucchini  
0                      No   No   No  No    No     No       No  
1                      No   No   No  No    No     No       No  
2                      No   No   No  No    No     No       No  
3                      No   No   No  No    No     No       No  
4                      No   No   No  No    No     No       No  

[5 rows x 384 columns]

Check the data dimensions

In [5]:
recipes.shape

(57691, 384)

Our data set have 57691 recipes with 384 ingredients.

#### Rename the `country` column to `Cuisine`

In [6]:
recipes.rename(columns={"country":"cuisine"}, inplace=True)
recipes

cuisine almond angelica anise anise_seed apple apple_brandy apricot  \
0      Vietnamese     No       No    No         No    No           No      No   
1      Vietnamese     No       No    No         No    No           No      No   
2      Vietnamese     No       No    No         No    No           No      No   
3      Vietnamese     No       No    No         No    No           No      No   
4      Vietnamese     No       No    No         No    No           No      No   
...           ...    ...      ...   ...        ...   ...          ...     ...   
57686       Japan     No       No    No         No    No           No      No   
57687       Japan     No       No    No         No    No           No      No   
57688       Japan     No       No    No         No    No           No      No   
57689       Japan     No       No    No         No    No           No      No   
57690       Japan     No       No    No         No    No           No      No   

      armagnac artemisia  ... whiskey white_bread white_wine  \
0           No        No  ...      No          No         No   
1           No        No  ...      No          No         No   
2           No        No  ...      No          No         No   
3           No        No  ...      No          No         No   
4           No        No  ...      No          No         No   
...        ...       ...  ...     ...         ...        ...   
57686       No        No  ...      No          No         No   
57687       No        No  ...      No          No         No   
57688       No        No  ...      No          No         No   
57689       No        No  ...      No          No         No   
57690       No        No  ...      No          No         No   

      whole_grain_wheat_flour wine wood yam yeast yogurt zucchini  
0                          No   No   No  No    No     No       No  
1                          No   No   No  No    No     No       No  
2                          No   No   No  No    No     No       No  
3                          No   No   No  No    No     No       No  
4                          No   No   No  No    No     No       No  
...                       ...  ...  ...  ..   ...    ...      ...  
57686                      No   No   No  No    No     No       No  
57687                      No   No   No  No    No     No       No  
57688                      No   No   No  No    No     No       No  
57689                      No   No   No  No    No     No       No  
57690                      No   No   No  No    No     No       No  

[57691 rows x 384 columns]

In [21]:
display(recipes["cuisine"].value_counts().index)
display(len(recipes["cuisine"].value_counts().index))

Index(['American', 'Mexico', 'Italian', 'Italy', 'Asian', 'French',
       'east_asian', 'Canada', 'korean', 'Mexican', 'western',
       'Southern_SoulFood', 'India', 'Jewish', 'Spanish_Portuguese',
       'Mediterranean', 'UK-and-Ireland', 'Indian', 'France', 'MiddleEastern',
       'Central_SouthAmerican', 'Germany', 'Eastern-Europe', 'Chinese',
       'Greek', 'English_Scottish', 'Caribbean', 'Thai', 'Scandinavia',
       'EasternEuropean_Russian', 'Cajun_Creole', 'Moroccan', 'Japanese',
       'China', 'Thailand', 'African', 'Southwestern', 'South-America',
       'japanese', 'Scandinavian', 'chinese', 'Irish', 'Japan', 'Spain',
       'italian', 'Vietnamese', 'North-African', 'German', 'Portugal',
       'Philippines', 'Korea', 'Netherlands', 'Lebanon', 'Vietnam', 'Austria',
       'Iran', 'Switzerland', 'Pakistan', 'Malaysia', 'asian', 'Turkey',
       'South-African', 'mexico', 'West-African', 'Indonesia', 'Belgium',
       'East-African', 'Israel', 'Bangladesh'],
      dtype='

69

There are duplicates in the cuisine columns that we need to take care of.

In [29]:
recipes["cuisine"] = recipes["cuisine"].apply(str.lower)
display(recipes["cuisine"].value_counts().index)
display(len(recipes["cuisine"].value_counts().index))

Index(['american', 'italian', 'mexico', 'italy', 'asian', 'french',
       'east_asian', 'canada', 'korean', 'mexican', 'western',
       'southern_soulfood', 'india', 'jewish', 'chinese', 'spanish_portuguese',
       'mediterranean', 'uk-and-ireland', 'indian', 'france', 'middleeastern',
       'central_southamerican', 'germany', 'japanese', 'eastern-europe',
       'greek', 'english_scottish', 'caribbean', 'thai', 'scandinavia',
       'easterneuropean_russian', 'cajun_creole', 'moroccan', 'china',
       'thailand', 'african', 'southwestern', 'south-america', 'scandinavian',
       'irish', 'japan', 'spain', 'vietnamese', 'north-african', 'german',
       'portugal', 'philippines', 'korea', 'netherlands', 'lebanon', 'vietnam',
       'austria', 'iran', 'switzerland', 'pakistan', 'malaysia',
       'south-african', 'turkey', 'west-african', 'indonesia', 'belgium',
       'east-african', 'israel', 'bangladesh'],
      dtype='object')

64

Remove duplicates

In [31]:
recipes.loc[recipes["cuisine"] == "austria", "cuisine"] = "austrian"
recipes.loc[recipes["cuisine"] == "belgium", "cuisine"] = "belgian"
recipes.loc[recipes["cuisine"] == "china", "cuisine"] = "chinese"
recipes.loc[recipes["cuisine"] == "canada", "cuisine"] = "canadian"
recipes.loc[recipes["cuisine"] == "netherlands", "cuisine"] = "dutch"
recipes.loc[recipes["cuisine"] == "france", "cuisine"] = "french"
recipes.loc[recipes["cuisine"] == "germany", "cuisine"] = "german"
recipes.loc[recipes["cuisine"] == "india", "cuisine"] = "indian"
recipes.loc[recipes["cuisine"] == "indonesia", "cuisine"] = "indonesian"
recipes.loc[recipes["cuisine"] == "iran", "cuisine"] = "iranian"
recipes.loc[recipes["cuisine"] == "italy", "cuisine"] = "italian"
recipes.loc[recipes["cuisine"] == "japan", "cuisine"] = "japanese"
recipes.loc[recipes["cuisine"] == "israel", "cuisine"] = "israeli"
recipes.loc[recipes["cuisine"] == "korea", "cuisine"] = "korean"
recipes.loc[recipes["cuisine"] == "lebanon", "cuisine"] = "lebanese"
recipes.loc[recipes["cuisine"] == "malaysia", "cuisine"] = "malaysian"
recipes.loc[recipes["cuisine"] == "mexico", "cuisine"] = "mexican"
recipes.loc[recipes["cuisine"] == "pakistan", "cuisine"] = "pakistani"
recipes.loc[recipes["cuisine"] == "philippines", "cuisine"] = "philippine"
recipes.loc[recipes["cuisine"] == "scandinavia", "cuisine"] = "scandinavian"
recipes.loc[recipes["cuisine"] == "spain", "cuisine"] = "spanish_portuguese"
recipes.loc[recipes["cuisine"] == "portugal", "cuisine"] = "spanish_portuguese"
recipes.loc[recipes["cuisine"] == "switzerland", "cuisine"] = "swiss"
recipes.loc[recipes["cuisine"] == "thailand", "cuisine"] = "thai"
recipes.loc[recipes["cuisine"] == "turkey", "cuisine"] = "turkish"
recipes.loc[recipes["cuisine"] == "vietnam", "cuisine"] = "vietnamese"
recipes.loc[recipes["cuisine"] == "uk-and-ireland", "cuisine"] = "uk-and-irish"
recipes.loc[recipes["cuisine"] == "irish", "cuisine"] = "uk-and-irish"

display(len(recipes["cuisine"].value_counts().index))

50

Now we have a total of **50** different cuisines instead of **69**.

Remove cuisine with less than **50** recipes.

In [60]:
to_remove = recipes["cuisine"].value_counts()[recipes["cuisine"].value_counts() < 50]
to_remove

philippine       43
dutch            32
lebanese         31
austrian         21
iranian          21
swiss            20
pakistani        19
malaysian        18
south-african    16
turkish          16
west-african     13
indonesian       12
east-african     11
belgian          11
israeli           9
bangladesh        4
Name: cuisine, dtype: int64

In [77]:
recipes = recipes[~recipes["cuisine"].isin(to_remove.index)].reset_index(drop=True)

In [78]:
recipes

cuisine almond angelica anise anise_seed apple apple_brandy apricot  \
0      vietnamese     No       No    No         No    No           No      No   
1      vietnamese     No       No    No         No    No           No      No   
2      vietnamese     No       No    No         No    No           No      No   
3      vietnamese     No       No    No         No    No           No      No   
4      vietnamese     No       No    No         No    No           No      No   
...           ...    ...      ...   ...        ...   ...          ...     ...   
57389    japanese     No       No    No         No    No           No      No   
57390    japanese     No       No    No         No    No           No      No   
57391    japanese     No       No    No         No    No           No      No   
57392    japanese     No       No    No         No    No           No      No   
57393    japanese     No       No    No         No    No           No      No   

      armagnac artemisia artichoke asparagus avocado bacon baked_potato balm  \
0           No        No        No        No      No    No           No   No   
1           No        No        No        No      No    No           No   No   
2           No        No        No        No      No    No           No   No   
3           No        No        No        No      No    No           No   No   
4           No        No        No        No      No    No           No   No   
...        ...       ...       ...       ...     ...   ...          ...  ...   
57389       No        No        No        No      No    No           No   No   
57390       No        No        No        No      No    No           No   No   
57391       No        No        No        No      No    No           No   No   
57392       No        No        No        No      No   Yes           No   No   
57393       No        No        No        No      No    No           No   No   

      banana barley bartlett_pear basil bay bean beech beef beef_broth  \
0         No     No            No   Yes  No   No    No   No         No   
1         No     No            No    No  No   No    No   No         No   
2         No     No            No    No  No   No    No   No         No   
3         No     No            No   Yes  No  Yes    No   No        Yes   
4         No     No            No    No  No   No    No   No         No   
...      ...    ...           ...   ...  ..  ...   ...  ...        ...   
57389     No     No            No    No  No   No    No   No         No   
57390     No     No            No    No  No   No    No   No         No   
57391     No    Yes            No    No  No   No    No   No         No   
57392     No     No            No    No  No   No    No   No         No   
57393     No    Yes            No    No  No   No    No   No         No   

      beef_liver beer beet bell_pepper bergamot berry bitter_orange  \
0             No   No   No          No       No    No            No   
1             No   No   No          No       No    No            No   
2             No   No   No          No       No    No            No   
3             No   No   No          No       No    No            No   
4             No   No   No          No       No    No            No   
...          ...  ...  ...         ...      ...   ...           ...   
57389         No   No   No          No       No    No            No   
57390         No   No   No          No       No    No            No   
57391         No   No   No          No       No    No            No   
57392         No   No   No          No       No    No            No   
57393         No   No   No          No       No    No            No   

      black_bean black_currant black_mustard_seed_oil black_pepper  \
0             No            No                     No           No   
1             No            No                     No          Yes   
2             No            No                     No           No   
3             No            No                     No 

We have removed **297** rows.
Now we need to convert `No` to `0` and `Yes` to `1`.

In [81]:
recipes = recipes.replace({"No":0, "Yes":1})
recipes.head()

cuisine  almond  angelica  anise  anise_seed  apple  apple_brandy  \
0  vietnamese       0         0      0           0      0             0   
1  vietnamese       0         0      0           0      0             0   
2  vietnamese       0         0      0           0      0             0   
3  vietnamese       0         0      0           0      0             0   
4  vietnamese       0         0      0           0      0             0   

   apricot  armagnac  artemisia  artichoke  asparagus  avocado  bacon  \
0        0         0          0          0          0        0      0   
1        0         0          0          0          0        0      0   
2        0         0          0          0          0        0      0   
3        0         0          0          0          0        0      0   
4        0         0          0          0          0        0      0   

   baked_potato  balm  banana  barley  bartlett_pear  basil  bay  bean  beech  \
0             0     0       0       0              0      1    0     0      0   
1             0     0       0       0              0      0    0     0      0   
2             0     0       0       0              0      0    0     0      0   
3             0     0       0       0              0      1    0     1      0   
4             0     0       0       0              0      0    0     0      0   

   beef  beef_broth  beef_liver  beer  beet  bell_pepper  bergamot  berry  \
0     0           0           0     0     0            0         0      0   
1     0           0           0     0     0            0         0      0   
2     0           0           0     0     0            0         0      0   
3     0           1           0     0     0            0         0      0   
4     0           0           0     0     0            0         0      0   

   bitter_orange  black_bean  black_currant  black_mustard_seed_oil  \
0              0           0              0                       0   
1              0           0              0                       0   
2              0           0              0                       0   
3              0           0              0                       0   
4              0           0              0                       0   

   black_pepper  black_raspberry  black_sesame_seed  black_tea  blackberry  \
0             0                0                  0          0           0   
1             1                0                  0          0           0   
2             0                0                  0          0           0   
3             0                0                  0          0           0   
4             0                0                  0          0           0   

   blackberry_brandy  blue_cheese  blueberry  bone_oil  bourbon_whiskey  \
0                  0            0          0         0                0   
1                  0            0          0         0                0   
2                  0            0          0         0                0   
3                  0            0          0         0                0   
4                  0            0          0         0                0   

   brandy  brassica  bread  broccoli  brown_rice  brussels_sprout  buckwheat  \
0       0         0      0         0           0                0          0   
1       0         0      0         0           0                0          0   
2       0         0      0         0           0                0          0   
3       0         0      0         0           0                0          0   
4       0         0      0         0           0                0          0   

   butter  buttermilk  cabbage  cabernet_sauvignon_wine  cacao  \
0       0           0        0                        0      0   
1       0           0        0                        0      0   
2       0           0        0                        0      0   
3       0           0        0                        0      0   
4       